sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -91,6 +91,14 @@ abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
   }
 
   /**
+   * Returns the maximum number of rows that this plan may compute.
+   *
+   * Any operator that a Limit can be pushed passed should override this function (e.g., Union).
+   * Any operator that can push through a Limit should override this function (e.g., Project).
+   */
+  def maxRows: Option[Expression] = None
```

Comment:
It's a `def` not a `lazy val`, I think it's fine with constant-folding rules.
BTW using `Option[Expression]` may be sub-optimal for something like `Union`, as its `maxRow: Some(children.flatMap(_.maxRows).reduce { (a, b) => Add(a, b) })` can't be constant-folded(`maxRow` is a method) and will always be a non-literal expression, we can't push limit through it.